### DISCUSSION TITLE
Perform pipeline order checking at build time

### DISCUSSION THREAD

**Discussion by robodobdob** [dotnet/aspnetcore#51615] [Ideas]
So often I’ve wondered why something isn’t working and it’s then I find some obscure document saying “make sure this goes between app.Something() and app.SomethingElse()”.

What I'd like to see with the whole pipeline thing is some sort of build time dependency check.

Could we have some sort of decoration on the pipeline middlewares like [NotBefore(…)] or [NotAfter(…)] so we can tell if a pipeline step isn't going to work as expected. Maybe it could also emit a severity. 

This would be an optional thing to only use if the middleware has these sort of dependencies.

**Comment by david-acker** [comment]
There are a handful of analyzers that emit diagnostics for issues with the registration order of things in the pipeline, for example [ASP0001: Authorization middleware is incorrectly configured](https://learn.microsoft.com/en-us/aspnet/core/diagnostics/asp0001?view=aspnetcore-7.0) or #48180 which will warn when a custom `IProblemDetailsWriter` is added in the wrong order.

But I agree, it would be beneficial to have checks in place for more of these scenarios, where feasible. I just noticed that a few of the middlewares listed [here](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/middleware/?view=aspnetcore-8.0&tabs=aspnetcore2x#built-in-middleware) (CORS, Output Caching, Response Caching, Request Timeouts) don't appear to have corresponding analyzers, although it seems like it would be relatively easy to create ones.

Are there any others that you've run into specifically?

**Reply by robodobdob** [reply]
I can't list specifically off the top of my head, but I have recently had a few 3rd-party middlewares that weren't functioning as expected because they were in the wrong order. I only found out they were wrong by Googling the packages and lucking on a post saying which order they had to run in.

I suppose this could be mitigated by the authors simply adding some Intellisense docs to their entry methods, so at least if I hover the method it could say "This must go between app.Something() and app.OtherThing()"
